### PR TITLE
Don't assume memory layout of std::net::SocketAddr

### LINKED ISSUE #105: net2 assumes the layout of std::net::SocketAddrV{4,6} matches libc::sockaddr
The problem code is here: https://github.com/deprecrated/net2-rs/blob/77a6eb4d5eb36adc81ee21ab7ffebc62fdf2ab88/src/socket.rs#L87-L142

It assumes that the layout of `std::net::SocketAddrV{4,6}` matches `libc::sockaddr`, but this can't be assumed.

Related issues: https://github.com/tokio-rs/mio/issues/1386, https://github.com/rust-lang/socket2-rs/issues/119, https://github.com/rust-lang/rust/pull/78802.

Mio has roughly the same code, this was fixed in https://github.com/tokio-rs/mio/pull/1388, for socket2: https://github.com/rust-lang/socket2-rs/pull/120.

### PR DESCRIPTION
**Pull request by faern**
Fixes #105 

This crate assumes the memory layout of `std::net::SocketAddr`, something it should not do. The layout might be correct now, but it's not part of the contract `std` makes to the outside world. Having commonly used crates like this one depend on unspecified things like that makes it harder for `std` to evolve.

This PR changes that by correctly converting between the `std` representation and the system representation instead of just doing pointer casting.

Same kind of fix as was recently done in https://github.com/rust-lang/socket2-rs/pull/120 and https://github.com/tokio-rs/mio/pull/1388

### CODE REVIEW

**Comment by faern** [comment]
All of these PRs look a bit different. Due to how the libraries are structured, but also because they have different minimum supported Rust versions etc. I'm assuming we have to keep the compatibility with Rust 1.21.0 for this fix. We want as many people as possible to be able to upgrade as smoothly as possible to get rid of the possible unsoundness this code had.

**Comment by faern** [comment]
Ok. Done. Ready for review I believe. Ping @pfmooney ?

For some context, see this PR: https://github.com/rust-lang/rust/pull/78802. And possibly this internals thread: https://internals.rust-lang.org/t/why-are-socketaddrv4-socketaddrv6-based-on-low-level-sockaddr-in-6/13321

**Review by Thomasdezeeuw — APPROVED** [review]
The same changes were applied to socket2 and Mio.

**Comment by faern** [comment]
I realize this PR still has a lot of transmutes that I aimed to get rid of. The other PRs have better solutions to this. I will fix this later today.

**Comment by Thomasdezeeuw** [comment]
> I realize this PR still has a lot of transmutes that I aimed to get rid of. The other PRs have better solutions to this. I will fix this later today.

I was looking at that but the `be_bytes` functions are Rust 1.32.0+, which is higher than the minimum supported version. But you can use `swap_bytes`.

**Comment by faern** [comment]
Oh yeah. That's why I moved over to `transmute`. I forgot. Too many very similar PRs this weekend :sweat_smile:. In another PR we temporarily had transmutes for other reasons that we got rid of. No this PR is good to go IMO.

I don't want any byte swapping. I just want to convert `u32->[u8;4]` and back without any swapping. I could also just do a pointer cast instead of a transmute. But I'll let the reviewer of this crate decide on that. The current implementation should work and be good enough IMO.

**Comment by faern** [comment]
I have verified that both `miow` and `mio` builds and completes their test suites when based on this version of `net2`. Both on latest stable Rust and on the Rust toolchain built from https://github.com/rust-lang/rust/pull/78802. So I'm pretty confident this is correctly implemented.

@pfmooney Would it be possible to have this reviewed (and published) :see_no_evil:? This crate is at the bottom of a pretty large dependency chain doing this invalid memory assumption. So things kind of have to start here.

File: src/socket.rs
Comment (pfmooney) [inline src/socket.rs]: What are your thoughts about splitting `addr2raw` into separate `cfg(windows)` and `cfg(unix)` variants, rather than interleaving like this?  It might be a little easier to follow, event though the `unix` version will still include the various specialization bits.

File: tests/all.rs
Comment (pfmooney) [inline tests/all.rs]: If we're going to hard-code the port, what about picking something a little higher (16384-32768) and not covered by an `/etc/services` listing?  It would give us a better chance of avoiding a clash with an existing thing, should the tests be run on a system with other running services.

File: tests/all.rs
Comment (pfmooney) [inline tests/all.rs]: Same as above, Re: port selection

File: src/socket.rs
Comment (pfmooney) [inline src/socket.rs]: nit: Skip the "And it can be initialized cleaner from Rust." sentence fragment

**Review by pfmooney — COMMENTED** [review]
Thanks for putting this together, and please pardon the review delay.

File: tests/all.rs
Comment (faern) [reply tests/all.rs]: Now when I test it it's actually quite likely to run into "Address already in use". Not because something else happens to listen there. But when running the tests multiple times and the socket does not get cleaned up fast enough.. So what do you think. Should we skip testing this?

File: src/socket.rs
Comment (pfmooney) [inline src/socket.rs]: Will the alignment requirements of `u32` be fulfilled by the array emitted from `ip().octets()`?  If that isn't guaranteed, is it really that expensive to do with safe rust?

File: tests/all.rs
Comment (pfmooney) [reply tests/all.rs]: Failing on repeat runs (due to lingering sockets) seems less than ideal.  Skipping it probably makes sense for now.

Comment (faern) [reply tests/all.rs]: I think we're fine without it. Nothing tested this was correct before either. And I'm pretty confident it's correct if I can make it pass the `mio` and `tokio` test suites.

File: src/socket.rs
Comment (faern) [reply src/socket.rs]: I'm not 100% sure if this is true for all platforms, but I think it should be. The alignment of `Ipv4Addr` is 4, which is the same as `u32`. So the alignment of the `[u8; 4]` should be correct unless I'm mistaken.

We could solve it safely by using the built in `Into<u32>` etc. And the compiler might be able to optimize it away, but if not it will do a byte swap twice.

Comment (pfmooney) [reply src/socket.rs]: I think avoiding the `unsafe` is worth the minuscule cost of a double byte-swap (especially considering it might be optimized away).

Comment (faern) [reply src/socket.rs]: Correctness is of course way more important than performance. So if we can't say for sure that this is correct I agree we should change to a safer solution.

File: src/socket.rs
Comment (faern) [reply src/socket.rs]: Done. I also splitted the v4 and v6 versions since the function was so long. I think this makes sense and puts most of the code at way lower indentation.

File: src/socket.rs
Comment (faern) [reply src/socket.rs]: Fixed. Less `unsafe` is always nice!

**Comment by faern** [comment]
@pfmooney There were a lot of pushes in a row there. But now I have addressed all the feedback and I have tested what's currently in this PR under `miow` and `mio 0.6` successfully.

**Comment by pfmooney** [comment]
> @pfmooney There were a lot of pushes in a row there. But now I have addressed all the feedback and I have tested what's currently in this PR under `miow` and `mio 0.6` successfully.

Great.  I'll run it through a few smoke tests of my own while waiting for @Thomasdezeeuw to approve of the updated change.

File: src/socket.rs
Comment (faern) [reply src/socket.rs]: Oops. I realize `Ipv4Addr::octets` of course return an owned `[u8; 4]` and not a reference. As such, the IP data is copied onto the stack and the alignment is just 1, not 4. Good we changed it.

**Review by Thomasdezeeuw — APPROVED** [review]
LGTM. @faern I'm amusing you've tried this branch with Mio and all tests passes?

**Comment by faern** [comment]
> @faern I'm amusing you've tried this branch with Mio and all tests passes?

Yes. On Windows 10 and Linux only though. But all tests pass. Both on a standard build of Rust and on the custom build where I change the memory layout of `SocketAddr`. So with this PR it indeed seems to not rely on the memory layout :)

**Comment by faern** [comment]
@pfmooney ping?

**Comment by pfmooney** [comment]
Merged as 49b43f277afb1caf5104fcbe02bef581f7444686.  Thanks

**Comment by pfmooney** [comment]
...and published as [0.2.36](https://crates.io/crates/net2/0.2.36).

**Comment by faern** [comment]
Yay! Thank you :tada:

**Pull request closed without merging**
